### PR TITLE
LPS-173502 Allow saving null values for numeric object fields

### PR DESCRIPTION
**Pull request by devpedrotavares**
Related issue: [LPS-173502](https://liferay.atlassian.net/browse/LPS-173502)

### CODE REVIEW

File: modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectEntryLocalServiceImpl.java [hunk 1]
```diff
@@ -3686,21 +3694,32 @@ else if (valueString.isEmpty()) {
 		}
 		else if (sqlType == Types.DECIMAL) {
 			if (Validator.isNull(String.valueOf(value))) {
-				value = BigDecimal.ZERO;
+				preparedStatement.setBigDecimal(index, null);
```
Comment (murilostodolni) [inline modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectEntryLocalServiceImpl.java:3697]: Couldn't we use preparedStatement.setNull here?

File: modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/service/test/ObjectEntryLocalServiceTest.java [hunk 2]
```diff
@@ -1111,7 +1111,7 @@ public void testAddOrUpdateObjectEntry() throws Exception {
 		_assertCount(1);
 
 		_assertObjectEntryValues(
-			21, values,
+			22, values,
```
Comment (murilostodolni) [inline modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/service/test/ObjectEntryLocalServiceTest.java:1114]: Why you changed the `expectedValuesSize` if you don't add a new entry?

File: modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/service/test/ObjectEntryLocalServiceTest.java [hunk 3]
```diff
@@ -1766,7 +1766,7 @@ public void testGetValuesList() throws Exception {
 
 		_assertCount(1);
 
-		_assertObjectEntryValues(27, values1, valuesList.get(0));
+		_assertObjectEntryValues(28, values1, valuesList.get(0));
```
Comment (murilostodolni) [inline modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/service/test/ObjectEntryLocalServiceTest.java:1769]: Why you changed the `expectedValuesSize` if you don't add a new entry?

**Review by murilostodolni — CHANGES_REQUESTED** [review]
Hey @devpedrotavares 

I finished my review. Let me know if you have questions.

File: modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectEntryLocalServiceImpl.java (hunk 1 above)
Comment (devpedrotavares) [reply modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectEntryLocalServiceImpl.java]: Yes! I just updated the PR

File: modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/service/test/ObjectEntryLocalServiceTest.java (hunk 2 above)
Comment (devpedrotavares) [reply modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/service/test/ObjectEntryLocalServiceTest.java:1114]: This was to fix an existing test failure. I believe a field was added to the `_objectDefinition` and the number of values was not updated.
I put these changes in a separate commit.

File: modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/service/test/ObjectEntryLocalServiceTest.java (hunk 3 above)
Comment (devpedrotavares) [reply modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/service/test/ObjectEntryLocalServiceTest.java:1769]: This was to fix an existing test failure. I believe a field was added to the `_objectDefinition` and the number of values was not updated.
I put these changes in a separate commit.

**Comment by devpedrotavares** [comment]
ci:test:sf

**Review by murilostodolni — APPROVED** [review]
LGTM 🎉

**Review by carolmariaabb — CHANGES_REQUESTED** [review]

File: modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectEntryLocalServiceImpl.java [hunk 4]
```diff
@@ -3686,21 +3694,32 @@ else if (valueString.isEmpty()) {
...
-			preparedStatement.setBigDecimal(
-				index,
-				new BigDecimal(_toPeriodSeparator(String.valueOf(value))));
 		}
 		else if (sqlType == Types.DOUBLE) {
-			preparedStatement.setDouble(
-				index,
-				GetterUtil.getDouble(
-					_toPeriodSeparator(String.valueOf(value))));
+			if (Validator.isNull(value)) {
```
[the remark this answers is not in the document]
Comment (devpedrotavares) [reply modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectEntryLocalServiceImpl.java]: Well noticed. I changed it to:  `if ((value == null) || StringPool.BLANK.equals(value))`  [here](https://github.com/liferay-objects/liferay-portal/pull/2755/commits/68f9f0633a3a2e18fee80ccac980857bf57e254f#:~:text=if%20((value%20%3D%3D%20null)%20%7C%7C%20StringPool.BLANK.equals(value))%20%7B).

File: modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectEntryLocalServiceImpl.java [hunk 5]
```diff
@@ -3635,7 +3638,12 @@ private void _setColumn(
 		throws Exception {
 
 		if (sqlType == Types.BIGINT) {
-			preparedStatement.setLong(index, GetterUtil.getLong(value));
+			if (Validator.isNull(value)) {
+				preparedStatement.setNull(index, Types.BIGINT);
+			}
+			else {
+				preparedStatement.setLong(index, GetterUtil.getLong(value));
+			}
```
[the remark this answers is not in the document]
Comment (devpedrotavares) [reply modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectEntryLocalServiceImpl.java]: I had this [idea](https://github.com/liferay-objects/liferay-portal/pull/2755/commits/68f9f0633a3a2e18fee80ccac980857bf57e254f). Is that okay?

File: modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectEntryLocalServiceImpl.java [hunk 6]
```diff
@@ -3059,6 +3059,10 @@ private Long _getStorageDLFolderId(
 	 */
 	private Object _getValue(Object object, int sqlType) throws SQLException {
 		if (sqlType == Types.BIGINT) {
+			if (object == null) {
+				return null;
+			}
+
```
[the remark this answers is not in the document]
Comment (devpedrotavares) [reply modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectEntryLocalServiceImpl.java]: I had this [idea](https://github.com/liferay-objects/liferay-portal/pull/2755/commits/68f9f0633a3a2e18fee80ccac980857bf57e254f). Is that okay?

**Comment by devpedrotavares** [comment]
ci:test:sf

**Review by carolmariaabb — APPROVED** [review]

**Comment by carolmariaabb** [comment]
ci:test:object

**Comment by carolmariaabb** [comment]
ci:test:relevant

**Comment by devpedrotavares** [comment]
ci:test:relevant

**Comment by devpedrotavares** [comment]
ci:test:object

**Comment by devpedrotavares** [comment]
ci:stop

**Comment by devpedrotavares** [comment]
ci:test:relevant

**Comment by devpedrotavares** [comment]
ci:test:object

**Comment by carolmariaabb** [comment]
ci:test:object-environments

**Comment by devpedrotavares** [comment]
ci:test:relevant

**Comment by devpedrotavares** [comment]
ci:test:object

**Comment by devpedrotavares** [comment]
ci:test:object-environments

**Comment by devpedrotavares** [comment]
ci:stop

**Comment by devpedrotavares** [comment]
ci:test:relevant

**Comment by devpedrotavares** [comment]
ci:test:object

**Comment by devpedrotavares** [comment]
ci:test:object-environments

**Comment by devpedrotavares** [comment]
Headless tests failing: [testray](https://testray.liferay.com/home/-/testray/runs/compare?compare=true&view=details&statusA=2&testrayRunIdB=2448517907&testrayRunIdA=2448946955&statusB=3)

**Comment by devpedrotavares** [comment]
ci:test:relevant

**Comment by devpedrotavares** [comment]
ci:test:object

**Comment by devpedrotavares** [comment]
ci:test:object-environments

**Comment by devpedrotavares** [comment]
Hi @carolmariaabb , I added this [commit](https://github.com/liferay-objects/liferay-portal/pull/2755/commits/efadfdc5078424392439cb8b03c6dff0d7489a79) to try and fix the headless side. I ran CI to check if that'll be enough

**Comment by carolmariaabb** [comment]
> Hi @carolmariaabb , I added this [commit](https://github.com/liferay-objects/liferay-portal/pull/2755/commits/efadfdc5078424392439cb8b03c6dff0d7489a79) to try and fix the headless side. I ran CI to check if that'll be enough

Please send a message to the headless team so they can take a look at your changes. Thanks!

**Comment by devpedrotavares** [comment]
ci:test:relevant

**Comment by devpedrotavares** [comment]
ci:test:object

**Comment by devpedrotavares** [comment]
ci:test:sf

**Comment by devpedrotavares** [comment]
> > Hi @carolmariaabb , I added this [commit](https://github.com/liferay-objects/liferay-portal/pull/2755/commits/efadfdc5078424392439cb8b03c6dff0d7489a79) to try and fix the headless side. I ran CI to check if that'll be enough
> 
> Please send a message to the headless team so they can take a look at your changes. Thanks!

I opened this [thread](https://liferay.slack.com/archives/C5E1CRLJY/p1698870391344709) with the headless team

**Comment by devpedrotavares** [comment]
ci:forward:force

**Pull request closed without merging**